modules/iothubdeviceping/src/iothubdeviceping.c

```diff
@@ -0,0 +1,738 @@
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_KEY_NAME == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_KEY == "NULL") ||
+        (((const IOTHUBDEVICEPING_CONFIG *)configuration)->EH_COMP_NAME == "NULL"))
+    {
+        LogError("invalid arg brokerHandle=%p, configuration=%p\r\n", brokerHandle, configuration);
+        result = NULL;
+    }
+    else
+    {
+        result = malloc(sizeof(IOTHUBDEVICEPING_HANDLE_DATA));
+        if (result == NULL)
+        {
+            LogError("malloc returned NULL\r\n");
+        }
+        else
+        {
+            result->config = (IOTHUBDEVICEPING_CONFIG *)clone_iothub_device_ping_config(configuration);
+            result->iotHubClientHandle = IoTHubClient_LL_CreateFromConnectionString(((const IOTHUBDEVICEPING_CONFIG *)configuration)->DeviceConnectionString, HTTP_Protocol);
+            if (result->iotHubClientHandle == NULL)
+            {
+                free(result);
+                result = NULL;
+                LogError("iotHubClientHandle returned NULL\r\n");
+            }
+            else
+            {
+                result->broker = brokerHandle;
+                result->lockHandle = Lock_Init();
+                if (result->lockHandle == NULL)
+                {
+                    free(result);
+                    result = NULL;
+                    LogError("unable to Lock_Init");
+                }
+                else
+                {
+                    if (ThreadAPI_Create(&result->threadHandle, device_ping_thread, result) != THREADAPI_OK)
+                    {
+                        (void)Lock_Deinit(result->lockHandle);
+                        free(result);
+                        result = NULL;
+                        LogError("failed to spawn a thread");
+                    }
+                    else
+                    {
+                        // all is fine apparently
+                    }
+                }
+            }
+        }
+    }
+    return result;
+}
+
+static void IoTHubDevicePing_Destroy(MODULE_HANDLE moduleHandle)
+{
+    IOTHUBDEVICEPING_HANDLE_DATA *handleData = moduleHandle;
+    int notUsed;
+    if (handleData->iotHubClientHandle != NULL)
+    {
+        IoTHubClient_LL_Destroy(handleData->iotHubClientHandle);
+    }
+    if (Lock(handleData->lockHandle) != LOCK_OK)
+    {
+        LogError("not able to Lock, still setting the thread to finish");
+    }
+    else
+    {
+        Unlock(handleData->lockHandle);
+    }
+    if (ThreadAPI_Join(handleData->threadHandle, &notUsed) != THREADAPI_OK)
+    {
+        LogError("unable to ThreadAPI_Join, still proceeding in _Destroy");
+    }
+    (void)Lock_Deinit(handleData->lockHandle);
+    free(handleData);
+}
+
+static void IoTHubDevicePing_Receive(MODULE_HANDLE moduleHandle, MESSAGE_HANDLE messageHandle)
+{
+    (void)moduleHandle;
+    (void)messageHandle;
+    // no action, IoTHubDevicePing is not interested in any messages
```

Comment:
the purpose for this sample is to verify the gateway infrastructure is set up correct. yes it can be done as a device sdk sample but it won't serve that purpose.